### PR TITLE
Fix a type issue in odeint

### PR DESCRIPTION
**Pull request by FlorianH-1QBit**
This PR fixes the issue described in https://github.com/google/jax/issues/8757, and modifies a test to catch it.

### CODE REVIEW

File: jax/experimental/ode.py
Comment (FlorianH-1QBit) [inline jax/experimental/ode.py:95]: `alpha` is used only in l112, where it is combined with the two times `t0` and `dt`. It should have the same dtype as the times, rather than the return value `f0`.

**Comment by jakevdp** [comment]
Thanks for the fix - once the CLA is resolved, I'll take a look.

**Comment by FlorianH-1QBit** [comment]
Sounds good - my company signed the CLA, I believe it's waiting for approval from Google's end now. I'll let you know once we receive that so you can re-run the workflow!

**Comment by FlorianH-1QBit** [comment]
Hi @jakevdp , happy new year! Could you please re-run the CLA workflow? Thanks.

**Comment by jakevdp** [comment]
I re-ran it; the CLA check is still failing. You'll need to follow the instructions at https://cla.developers.google.com/ to register the email address you use to sign your commits.

**Comment by FlorianH-1QBit** [comment]
Sorry about that, something went wrong on our end. I'll ping you again when it's fixed.

File: jax/experimental/ode.py
Comment (jakevdp) [reply jax/experimental/ode.py:95]: Good catch – though what happens if someone inadvertently passes in an integer sequence of times? Should we promote `t0` and `dt` to inexact somewhere?

**Comment by jakevdp** [comment]
Thanks! Looks like the CLA check passes now.

File: jax/experimental/ode.py
Comment (FlorianH-1QBit) [reply jax/experimental/ode.py:95]: Good point - if an integer `t0` ever makes it here, `alpha` would be wrong if using `t0.dtype`!
Safer would be `f0.real.dtype` or `_complex_elem_type(f0)` from `jax._src.numpy.lax_numpy`?

Right now, `odeint(func=lambda x, t: x, y0=jnp.array(1.), t=jnp.array([0, 1]))` fails elsewhere, so if we want to make that run then `t` should probably be promoted further outside - in `_odeint` maybe? I don't know enough about jax types to do that unfortunately.

Comment (jakevdp) [reply jax/experimental/ode.py:95]: Makes sense – I think the easiest fix then would be to (somewhere) add `assert np.issubdtype(t, np.floating)`. The `odeint` docstring says that `t` is an array of floats, so there's probably not any good reason to handle other cases.

File: jax/experimental/ode.py
Comment (jakevdp) [inline jax/experimental/ode.py:173]:
```suggestion
    raise TypeError(f"t must be an array of floats, but got {t}.")
```

**Comment by jakevdp** [comment]
I think the errors are ones that have been fixed on the main brach. Can you rebase?

**Comment by jakevdp** [comment]
Also, we should squash all changes into a single commit before merging (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests)

**Comment by FlorianH-1QBit** [comment]
> I think the errors are ones that have been fixed on the main brach. Can you rebase?

Do you mean that the main branch uses f-strings for the `TypeError` messages? 
Right now the `*args` TypeError message also uses `.format` so I'll change that as well.

**Comment by jakevdp** [comment]
No, I meant the failed documentation build in the github CI. If you rebase on main, those errors should go away.

**Comment by jakevdp** [comment]
The `.format()` calls are leftover from older versions of Python. We generally prefer f strings now.

**Review by jakevdp — APPROVED** [review]
Looks great!

**Pull request merged**
